{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Reading Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def head(file_name):\n",
    "    \"\"\"\n",
    "    Prints the FIRST 10 lines of a file\n",
    "    :param file_name: a string\n",
    "    :return: None\n",
    "    \"\"\"\n",
    "    f = open(file_name,\"r\") \n",
    "    lineList = f.readlines()\n",
    "    f.close()\n",
    "    s1 = \"\".join(lineList[0:10])\n",
    "    print(s1)\n",
    "    return\n",
    "\n",
    "def tail(file_name):\n",
    "    \"\"\"\n",
    "    Prints the LAST 10 lines of a file\n",
    "    :param file_name: a string\n",
    "    :return: None\n",
    "    \"\"\"\n",
    "    f = open(file_name,\"r\") \n",
    "    lineList = f.readlines()\n",
    "    f.close()\n",
    "    s1 = \"\".join(lineList[len(lineList)-10:])\n",
    "    print(s1)\n",
    "    return\n",
    "\n",
    "def print_even(file_name):\n",
    "    \"\"\"\n",
    "    Prints the even numbered lines of a file\n",
    "    :param file_name: a string\n",
    "    :return: None\n",
    "    \"\"\"\n",
    "    f = open(file_name,\"r\") \n",
    "    lineList = f.readlines()\n",
    "    f.close()\n",
    "    s1 = \"\".join(lineList[1::2])\n",
    "    print(s1)\n",
    "    return\n",
    "\n",
    "def csv_list(file_name):\n",
    "    \"\"\"\n",
    "    Read in a CSV file to a 2D array (In python it is a list of lists)\n",
    "    :param file_name: a string\n",
    "    :return: a list of lists\n",
    "    \"\"\"\n",
    "    my_array=[]\n",
    "    with open(file_name) as myfile:\n",
    "        for line in myfile:\n",
    "            row = line.split(',')\n",
    "            my_array.append(row)\n",
    "    return my_array\n",
    "\n",
    "def get_csv_column(file_name, column):\n",
    "    \"\"\"\n",
    "    Reads in a CSV file and returns a list of values belonging to the column specified\n",
    "    :param file_name: a string\n",
    "    :param column: a positive integer\n",
    "    :return: a list\n",
    "    \"\"\"\n",
    "    my_array=[]\n",
    "    with open(file_name) as myfile:\n",
    "        for line in myfile:\n",
    "            row = line.split(',')\n",
    "            if(len(row) < column):\n",
    "                print('Column {} does not exist in row {}'.format(column - 1, row))\n",
    "            else:\n",
    "                my_array.append(row[column - 1])\n",
    "    return my_array\n",
    "\n",
    "def fasta_seqs(file_name):\n",
    "    \"\"\"\n",
    "    Reads in a FASTA file and returns a list of only the sequences\n",
    "    :param file_name: a string\n",
    "    :return: a list of strings\n",
    "    \"\"\"\n",
    "    seq_list = []\n",
    "    with open(file_name, 'r') as infile:\n",
    "        text = infile.read()\n",
    "        seqs = text.split('>')\n",
    "        for seq in seqs:\n",
    "            try:\n",
    "                x = seq.split('\\n', 1)\n",
    "                header = x[0]\n",
    "                sequence = x[1].replace('\\n', '')\n",
    "                seq_list.append(sequence)\n",
    "                \n",
    "            except:\n",
    "                pass\n",
    "    return seq_list\n",
    "\n",
    "def fasta_headers(file_name):\n",
    "    \"\"\"\n",
    "    Reads in a FASTA file and returns a list of only the headers (Lines that start with \">\")\n",
    "    :param file_name: a string\n",
    "    :return: a list of strings\n",
    "    \"\"\"\n",
    "    header_list = []\n",
    "    with open(file_name, 'r') as infile:\n",
    "        text = infile.read()\n",
    "        seqs = text.split('>')\n",
    "        for seq in seqs:\n",
    "            try:\n",
    "                x = seq.split('\\n', 1)\n",
    "                header = x[0]\n",
    "                sequence = x[1].replace('\\n', '')\n",
    "                header_list.append(header)\n",
    "                \n",
    "            except:\n",
    "                pass\n",
    "    return header_list\n",
    "\n",
    "\n",
    "def fasta_dict(file_name):\n",
    "    \"\"\"\n",
    "    Reads in a FASTA file and returns a dictionary of the format {header: sequence, ...}, where\n",
    "    the sequence headers are keys and the sequence is the value\n",
    "    :param file_name: a string\n",
    "    :return: a dictionary\n",
    "    \"\"\"\n",
    "    my_dict ={}\n",
    "    with open(file_name, 'r') as infile:\n",
    "        text = infile.read()\n",
    "        seqs = text.split('>')\n",
    "        for seq in seqs:\n",
    "            try:\n",
    "                x = seq.split('\\n', 1)\n",
    "                header = x[0]\n",
    "                sequence = x[1].replace('\\n', '')\n",
    "                my_dict[header] = sequence\n",
    "            except:\n",
    "                pass\n",
    "    return my_dict\n",
    "\n",
    "def fastq_to_fasta(file_name, new_name=None):\n",
    "    \"\"\"\n",
    "    Reads in a FASTQ file and writes it to a new FASTA file. This definition should also\n",
    "    keep the same file name and change the extension to from .fastq to .fasta if new_name is not specified.\n",
    "    EX: fastq_to_fasta('ecoli.fastq') should write to a new file called ecoli.fasta\n",
    "    :param file_name: a string\n",
    "    :param new_name: a string\n",
    "    :return: None\n",
    "    \"\"\"\n",
    "    oFile = file_name.split('.')\n",
    "    oFile[0] += '.fasta'\n",
    "    output = open(oFile[0],'w')\n",
    "    with open(file_name, 'r') as infile:\n",
    "        text = infile.read()\n",
    "        seqs = text.split('@')\n",
    "        for seq in seqs:\n",
    "            try:\n",
    "                x = seq.split('\\n',1)\n",
    "                header = x[0]\n",
    "                seq1 = x[1].replace('\\n', '')\n",
    "                sequence = seq1.split('+')\n",
    "                output.write('<'+header+ '\\n' + sequence[0] + '\\n')\n",
    "            except:\n",
    "                pass\n",
    "    output.close()    \n",
    "    return\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
